File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
# 프로그래머스 43163 단어 변환 LV.3
from collections import deque


def solution(begin, target, words):
    def bfs(start):
        q = deque()
        q.append([start, 0])
        while q:
            current, step = q.popleft()
            if current == target:
                return step

            for word in words:
                word_cnt = 0
                for i in range(len(word)):
                    if current[i] != word[i]:
                        word_cnt += 1

                if word_cnt == 1:
                    q.append([word, step + 1])

    if target not in words:
        return 0

    return bfs(begin)

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Word {
    String original;
    int cnt;
    
    Word(String original, int cnt) {
        this.original = original;
        this.cnt = cnt;
    }
}

class Solution {
    final Set<String> wordSet = new HashSet<>();
    
    public int solution(String begin, String target, String[] words) {
        for (String word : words) wordSet.add(word);
        if (!wordSet.contains(target)) return 0;
        
        return bfs(begin, target);
    }
    
    private int bfs(String begin, String target) {
        Queue<Word> q = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        q.add(new Word(begin, 0));
        visited.add(begin);
        
        int cnt = Integer.MAX_VALUE;
        while (!q.isEmpty()) {
            Word now = q.poll();
            
            if (now.original.equals(target)) cnt = Math.min(cnt, now.cnt);
            
            for (int idx = 0; idx < now.original.length(); idx++) {
                char[] cArr = now.original.toCharArray();
                for (int j = 0; j <= 26; j++) {
                    cArr[idx] = (char) ((int) 'a' + j);
                    String tmp = String.valueOf(cArr);
                    if (!visited.contains(tmp) && wordSet.contains(tmp)) {
                        q.add(new Word(tmp, now.cnt + 1));
                        visited.add(tmp);
                    }
                }
            }
        }
        
        return cnt;
    }
}